src/charm.py

```diff
@@ -240,6 +258,57 @@ def _on_get_admin_secrets_action(self, event: ActionEvent):
 
         event.set_results({"password": password if password else "", "chain": chain})
 
+    def _on_start_service_action(self, event: ActionEvent):
+        """Start the OpenSearch service from an action event."""
+        if self.opensearch.is_node_up():
+            event.set_results({"message": "OpenSearch is already started in this node."})
+            return
+
+        if not self._start_opensearch():
+            event.defer()
+            event.set_results(
+                {"message": "Something happened, the OpenSearch service will re-attempt a start."}
+            )
+            return
+
+        event.set_results({"message": "The OpenSearch service is attempting a start..."})
+
+    def _on_restart_service_action(self, event: ActionEvent):
+        """Restart the OpenSearch service from an action event."""
+        try:
+            self.unit.status = MaintenanceStatus(ServiceIsStopping)
+            self.opensearch.stop()
+            self.unit.status = MaintenanceStatus(ServiceStopped)
+        except OpenSearchStopError as e:
+            logger.error(e)
+            event.set_results(
+                {"message": "An error occurred during the stop of the OpenSearch service."}
+            )
+            self.unit.status = BlockedStatus(ServiceStopFailed)
+            return
+
+        if not self._start_opensearch():
+            event.defer()
+            event.set_results(
+                {"message": "Something happened, the OpenSearch service will re-attempt a start."}
+            )
+            return
+
+        event.set_results({"message": "The OpenSearch service is attempting a restart..."})
+
+    def _on_stop_service_action(self, event: ActionEvent):
+        """Stop the OpenSearch service from an action event."""
+        try:
+            self.unit.status = MaintenanceStatus(ServiceIsStopping)
+            self.opensearch.stop()
+            self.unit.status = WaitingStatus(ServiceStopped)
+            event.set_results({"message": "The OpenSearch service is stopping..."})
```

Comment:
Should the message that is sent as result of the event match the waiting status (`The OpenSearch service stopped.`)?